middleware/hosts/hosts.go

```diff
@@ -0,0 +1,178 @@
+package hosts
+
+import (
+	"errors"
+	"net"
+
+	"golang.org/x/net/context"
+
+	"github.com/coredns/coredns/middleware"
+	"github.com/coredns/coredns/middleware/pkg/dnsutil"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+)
+
+// Hosts is the middleware handler
+type Hosts struct {
+	Next middleware.Handler
+	*Hostsfile
+
+	Origins     []string
+	Fallthrough bool
+}
+
+// ServeDNS implements the middleware.Handle interface.
+func (h Hosts) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	if state.QClass() != dns.ClassINET {
+		return dns.RcodeServerFailure, middleware.Error(h.Name(), errors.New("can only deal with ClassINET"))
+	}
+	qname := state.Name()
+
+	answers := []dns.RR{}
+
+	// Precheck with the origins, i.e. are we allowed to looks here.
+	if h.Origins != nil {
```

Comment:
I'm also having trouble understanding, why there is precheck? Most of this code is repeated below?